src/filesystem.cc

```diff
@@ -1941,37 +2166,244 @@ GetFileSystem(const std::string& path, FileSystem** file_system)
+      const Status& match_status =
+          GetLongestMatchingCredential<std::pair<std::string, std::string>>(
+              as_credential, "as://", path.substr(5, std::string::npos),
+              credential_);
+      if (!match_status.IsOk()) {
+        return ReturnErrorOrReload(
+            load_status, match_status, path, file_system);
+      }
+      as_fs = new ASFileSystem(
+          path, credential_.first.c_str(), credential_.second.c_str());
+      const Status& client_status = as_fs->CheckClient();
+      if (!client_status.IsOk()) {
+        return ReturnErrorOrReload(
+            load_status, client_status, path, file_system);
+      }
+      *file_system = as_fs;
+      return Status::Success;
+    } else if (load_status.Message() == "Use legacy credential") {
+      as_fs = new ASFileSystem(path);
+      RETURN_IF_ERROR(as_fs->CheckClient());
+      *file_system = as_fs;
+      return Status::Success;
+    }
+    return load_status;
 #endif  // TRITON_ENABLE_AZURE_STORAGE
   }
 
   // Assume path is for local filesystem
-  static LocalFileSystem local_fs;
-  *file_system = &local_fs;
-
+  *file_system = local_fs = new LocalFileSystem();
   return Status::Success;
 }
 
 Status
-GetFileSystem(FileSystemType type, FileSystem** file_system)
+FileSystemManager::GetFileSystem(FileSystemType type, FileSystem** file_system)
 {
-  // FIXME currently this function only work for LOCAL, GCS because their
-  // construction is not path-dependent. And in my opinion here should be
-  // where the fs instances are placed and GetFileSystem() should call this
-  // function after it identifies the fs type from path.
-  switch (type) {
-    case FileSystemType::LOCAL:
-      return GetFileSystem("", file_system);
-    case FileSystemType::GCS:
-      return GetFileSystem("gs://", file_system);
-    default:
-      return Status(
-          Status::Code::UNSUPPORTED,
-          "The requested filesteam can not be accessed by type");
+  const Status& status_ = LoadCredential();
+  if (status_.IsOk()) {
+    // with credential file, use default credential (one with shortest name)
+    switch (type) {
+      case FileSystemType::LOCAL:
+        return GetFileSystem("", file_system);
+      case FileSystemType::GCS:
+        return GetFileSystem("gs://", file_system);
+      case FileSystemType::S3:
+        return GetFileSystem("s3://", file_system);
+      case FileSystemType::AS:
+        return GetFileSystem("as://", file_system);
+      default:
+        return Status(
+            Status::Code::UNSUPPORTED,
+            "The requested filesteam cannot be accessed by type");
+    }
+  } else if (status_.Message() == "Use legacy credential") {
+    // with env var, only LOCAL and GCS are not path-dependent
+    switch (type) {
+      case FileSystemType::LOCAL:
+        return GetFileSystem("", file_system);
+      case FileSystemType::GCS:
+        return GetFileSystem("gs://", file_system);
+      default:
```

Comment:
Maybe return specific error messages here on S3/Azure for clearer errors

Reply:
Yes, good suggestion. Updated.